File: assignCookies.py
from typing import List

"""
	Intuition:  Sort both list of children and cookies.  Eliminate the cookies that
				dont satisfy any child apetite.  Count those that do satisfy their appetite
			    and move on to next child.  The program should end as soon as either we
				have finish with cookies or children.

	RunTime: O(N Log N)
	Space Complexity: O(1)
"""

class Solution:
    def findContentChildren(self, g: List[int], s: List[int]) -> int:
        g.sort()
        s.sort()
     
        counter = 0;
        childPtr = 0;
        
        for i in range(0, len(s)):
            if (s[i] < g[childPtr]):
                continue;
            
            elif (s[i] >= g[childPtr]):
                counter += 1
                childPtr += 1
                
                if (childPtr >= len(g)):
                    break;
                
        return counter;


g = [1,2]
s = [1,2,3]

myObj = Solution()
answer = myObj.findContentChildren(g, s)
print(answer)

File: assignCookies.cpp
#include <iostream>
using namespace std;

#include <vector>
#include <algorithm>

/*
	Intuition: Sort both arrays.  Traverse through the cookies array
			   trying to get rid of cookies that are smaller than childrens'
			   appetite.  When we find a cookie that satisfy a child then we increase counter
			   and move to another kid.  If we have finished all children before finishing
			   all cookies, we simply break and end the counting.

	RunTime: O(N Log N)
	Space Complexity: O(1)
*/



class Solution {
public:
    int findContentChildren(vector<int>& g, vector<int>& s) 
    {
        sort(g.begin(), g.end());
        sort(s.begin(), s.end());
        
        int counter = 0;
        int childPtr = 0;
        
        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] < g[childPtr])
            {
                continue;
            }
            
            else if (s[i] >= g[childPtr])
            {
                counter++;
                childPtr++;
                
                if (childPtr >= g.size())
                {
                    break;
                }
            }
        }
        
        return counter;
    }
};

int main()
{
	vector<int> g = {1,2};
	vector<int> s = {1,2,3};

	Solution myObj = Solution();
	
	int answer = myObj.findContentChildren(g, s);

	cout << answer << endl;

	return 0;
}
